scheduler/executor/workflow_executor.go

```diff
@@ -0,0 +1,349 @@
+	rateLimiter          ratelimit.Limiter
+	metrics              schedulerMetrics
+	adminServiceClient service.AdminServiceClient
+}
+
+func (w *workflowExecutor) CheckPointState(ctx context.Context) {
+	defer func() {
+		if err := recover(); err != nil {
+			w.metrics.CheckPointPanicCounter.Inc()
+			logger.Fatalf(context.Background(), fmt.Sprintf("caught panic: %v [%+v]", err, string(debug.Stack())))
+		}
+	}()
+
+	for true {
+		var bytesArray []byte
+		f := bytes.NewBuffer(bytesArray)
+		// Only write if the snapshot has contents and not equal to the previous snapshot
+		if !w.snapshot.IsEmpty() {
+			err := w.snapShotReaderWriter.WriteSnapshot(f, w.snapshot)
+			// Just log the error
+			if err != nil {
+				w.metrics.CheckPointCreationErrCounter.Inc()
+				logger.Errorf(ctx, "unable to write the snapshot to buffer due to %v", err)
+			}
+			err = w.db.ScheduleEntitiesSnapshotRepo().CreateSnapShot(ctx, models.ScheduleEntitiesSnapshot{
+				Snapshot: f.Bytes(),
+			})
+			if err != nil {
+				w.metrics.CheckPointSaveErrCounter.Inc()
+				logger.Errorf(ctx, "unable to save the snapshot to the database due to %v", err)
+			}
+		}
+		time.Sleep(snapshotWriterSleepTime * time.Second)
+	}
+}
+
+func (w *workflowExecutor) CatchUpAllSchedules(ctx context.Context, schedules []models.SchedulableEntity, toTime time.Time) error {
+	logger.Debugf(ctx, "catching up [%v] schedules until time %v", len(schedules), toTime)
+	for _, s := range schedules {
+		fromTime := time.Now()
+		// If the schedule is not active, don't do anything else use the updateAt timestamp to find when the schedule became active
+		// We support catchup only from the last active state
+		// i.e if the schedule was Active(t1)-Archive(t2)-Active(t3)-Archive(t4)-Active-(t5)
+		// And if the scheduler was down during t1-t5 , then when it comes back up it would use t5 timestamp
+		// to catch up until the current timestamp
+		// Here the assumption is updateAt timestamp changes for active/inactive transitions and no other changes.
+		if !*s.Active {
+			logger.Debugf(ctx, "schedule %+v was inactive during catchup", s)
+			continue
+		} else {
+			fromTime = s.UpdatedAt
+		}
+		nameOfSchedule := GetScheduleName(s)
+		lastT := w.snapshot.GetLastExecutionTime(nameOfSchedule)
+		if !lastT.IsZero() && lastT.After(s.UpdatedAt) {
+			fromTime = lastT
+		}
+		logger.Debugf(ctx, "catching up schedule %+v from %v to %v", s, fromTime, toTime)
+		err := w.CatchUpSingleSchedule(ctx, s, fromTime, toTime)
+		if err != nil {
+			return err
+		}
+		logger.Debugf(ctx, "caught up successfully on the schedule %+v from %v to %v", s, fromTime, toTime)
+	}
+	return nil
+}
+
+func (w *workflowExecutor) CatchUpSingleSchedule(ctx context.Context, s models.SchedulableEntity, fromTime time.Time, toTime time.Time) error {
+	nameOfSchedule := GetScheduleName(s)
+	var catchUpTimes []time.Time
+	var err error
+	catchUpTimes, err = w.goGfInterface.GetCatchUpTimes(s, fromTime, toTime)
```

Comment:
this is not goGCInterface anymore right?

Reply:
renamed from goGfInterface  -> goCronInterface
We are still using the interface to get a slice of catch up times.
Do you have any concerns here?